src/backend/xilinx/toplevel.rs

```diff
@@ -79,13 +98,39 @@ impl Backend for XilinxInterfaceBackend {
     }
 }
 
-fn external_memories(comp: &ir::Component) -> Vec<String> {
-    // find external memories
+// XXX(nathanielnrn) I __think__ this should always be in the same order because
+// of IdList being in the same order
+// also, should this be a macro?
```

Comment:
What should be a macro? This function? Is so why?

Reply:
Initially I was wondering if the helper function should be a macro due to some trouble I was having implementing it. But if the helper function looks ok:
```
fn external_memories_cells(
    comp: &ir::Component,
) -> Vec<calyx::ir::RRC<ir::Cell>> {
    comp.cells
        .iter()
        .filter(|cell_ref| {
            matches!(cell_ref.borrow().get_attribute("external"), Some(&1))
        // find external memories
        .filter(|cell_ref| cell_ref.borrow().attributes.has("external"))
        .cloned()
        .collect()
}
```
And the way it is used in these two functions is fine:

```
// Returns a vector of tuples containing external memory info of form:
// (WIDTH, SIZE, IDX_SIZE)
fn get_mem_info(comp: &ir::Component) -> Vec<(u64, u64, u64)> {
    external_memories_cells(comp)
        .iter()
        .map(|cell_ref| {
            (
                cell_ref.borrow().get_parameter("WIDTH").unwrap(),
                cell_ref.borrow().get_parameter("SIZE").unwrap(),
                cell_ref.borrow().get_parameter("IDX_SIZE").unwrap(),
            )
        })
        .collect()
}

// Returns Vec<String> of memory names
fn external_memories(comp: &ir::Component) -> Vec<String> {
    external_memories_cells(comp)
        .iter()
        .map(|cell_ref| cell_ref.borrow().name().to_string())
        .collect()
}
```

Then it should be fine to leave it as is.